### PR TITLE
lion-calendar

### PR DESCRIPTION
**Pull request by bashmish**
Reusable calendar component 📆

### CODE REVIEW

**Comment by tlouisse** [comment]
Looks really good.

Few questions/suggestions:
- rename 'centralDate' to 'focusableDate', since it's more clear from the name what it does.
- why are 'goToNextMonth' and 'goToPreviousMonth' public methods? Are they supposed to be called by Application Developers? If not, protected would be better

Naming conventions for Subclassers:
- '__renderHeader' instead of '_headerTemplate' or 'headerTemplate'. I would opt for '_headerTemplate' (aligned with existing naming convention, one underscore since available for subclassers)

File: packages/calendar/package.json [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  "publishConfig": {
+    "access": "public"
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/ing-bank/lion.git",
+    "directory": "packages/calendar"
+  },
+  "scripts": {
+    "prepublishOnly": "../../scripts/insert-header.js"
```
Comment (daKmoR) [inline packages/calendar/package.json]:
```suggestion
    "prepublishOnly": "../../scripts/npm-prepublish.js"
```

prepublish script changed

**Comment by erikkroes** [comment]
Following is a list of accessibility issues found.

## General

- Screen reader navigation doesn't work in all AT/browser combinations. Adding `role="application"` on the component fixes this issue.
- `aria-selected="true" is not announced. Adding an sr-only text "selected" would fix this.
- Min and max date are not clear for SR users. Suggested fix: if navigating beyond min or max, put focus on min or max. Add sr-only text: "First selectable date" and "Last selectable date" or similar to the min and max date.
- When there are disabled dates, these are skipped by keyboard navigation. This should not be the case as it is navigation for the grid, not just the interactive elements. Preferably, navigation should put the focus for cells with disabled buttons on the cell.

##  VO + Safari

- After navigating to another month, the date gets announced as it should. After a short pause it is followed by "2 <month-name>". The month name is most likely from the `aria-live`, as it should be. The 2 should not be announced.

## NVDA + Firefox

- When navigating to another month, sometimes another day than the navigation target gets pronounced first. This day is in the same row and column as the one you're coming from. As if it changes the month first, with the focus still on the same row and column, before putting the focus on the right cell.
- The string that is the accessible name of the button gets pronounced twice. First as the accessible name of the gridcell, then for the button. This seems to be default browser/AT behaviour, but if it can be removed, that would improve the user experience.

File: packages/calendar/package.json (hunk 1 above)
Comment (bashmish) [reply packages/calendar/package.json]: thanks!

**Comment by bashmish** [comment]
@erikkroes I added `role="application"`. And I agree with all other points as well, but they are probably for later. Some require translations for all currently supported languages, other involve quite some rework and rethinking even (especially the one where we want to enable days for keyboard navigation while disabling the buttons behind them). Let's create individual bug report and feature requests after this is released.

**Comment by bashmish** [comment]
> rename 'centralDate' to 'focusableDate', since it's more clear from the name what it does.

This date has more behavior than just make a day focusable. For example it will change the view if you provide a new value from a different month. Also if I would reverse the meaning of `focusable` applying to other days I can assume that all other days will be disabled, while in fact it does not disable anything, but just gives more priority to one of them. All in all I thing `centralDate` is a good compromise between having a good name with unexpected behavior and a bad name without any expectation (which you can read about in the documentation).

File: packages/localize/src/date/getMonthNames.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { normalizeDate } from './normalizeDate.js';
+
+const monthsLocaleCache = {};
+
+export function getMonthNames({ locale }) {
```
Comment (tlouisse) [inline packages/localize/src/date/getMonthNames.js]: Did we deliberately choose not to make notation here configurable?

**Comment by bashmish** [comment]
> why are 'goToNextMonth' and 'goToPreviousMonth' public methods? Are they supposed to be called by Application Developers? If not, protected would be better

They were protected and used in our extensions. But then I realised that programmatically changing months can be quite useful and their API will likely never change due to simplicity of the interface and unambiguity. And I made them public.

**Comment by bashmish** [comment]
> 'dayPreprocessor' would be called 'dayProcessor' I thought? It would be preprocessing in the sense that it runs before your render, post processing in the sense that it runs after the default processing. Both are in a level of abstraction that the Application Developer should not be bothered with. He only wants to process his dates, so that he can add some extra meta info in his calendar

`pre` gives me a hint that it happens before smth. As a user I can then learn that this happens before the rendering. Does it really matter for the end user? Probably not. But does not make harm either. `processor` is not a perfect name in the first place, but we could find any better :(

**Comment by tlouisse** [comment]
> > why are 'goToNextMonth' and 'goToPreviousMonth' public methods? Are they supposed to be called by Application Developers? If not, protected would be better
> 
> They were protected and used in our extensions. But then I realised that programmatically changing months can be quite useful and their API will likely never change due to simplicity of the interface and unambiguity. And I made them public.

Could make sense indeed, for instance when you create your own calendar overlay and want to add some interaction from outer layer.

**Comment by tlouisse** [comment]
> > 'dayPreprocessor' would be called 'dayProcessor' I thought? It would be preprocessing in the sense that it runs before your render, post processing in the sense that it runs after the default processing. Both are in a level of abstraction that the Application Developer should not be bothered with. He only wants to process his dates, so that he can add some extra meta info in his calendar
> 
> `pre` gives me a hint that it happens before smth. As a user I can then learn that this happens before the rendering. Does it really matter for the end user? Probably not. But does not make harm either. `processor` is not a perfect name in the first place, but we could find any better :(

The less implementation details we expose, the better. Maybe preprocessor and processor are equally vague:)  Can't come up with a better name, though. Good docs and an example would be sufficient, then. Same for 'centralDate' (clear description in static 'get properties')

**Comment by bashmish** [comment]
> '__renderHeader' instead of '_headerTemplate' or 'headerTemplate'. I would opt for '_headerTemplate' (aligned with existing naming convention, one underscore since available for subclassers)

We never had a situation with reusable templates in the past in `lion`. Calendar is the first component where we have both reusable templates (I suggest to call them just `template functions`) which receive everything via arguments and non-reusable templates bound to the context of a web component (I suggest to call them `render functions`).

e.g. `__renderData` is a render function which uses the template `dataTemplate`

https://github.com/ing-bank/lion/pull/37/files#diff-0d8b3c4eac421dbadc591e2ed5b80f6dR280

Therefore I had to introduce a new convention here. And I don't think it is a good idea to follow the previous approach with no separation between those different things. Or it will be like:

```js
 __dataTemplate() {
    return dataTemplate(this.__data, {
      monthsLabels: getMonthNames({ locale: this.__getLocale() }),
      weekdaysShort: getWeekdayNames({
        locale: this.__getLocale(),
        style: this.weekdayHeaderNotation,
        firstDayOfWeek: this.firstDayOfWeek,
      }),
      weekdays: getWeekdayNames({
        locale: this.__getLocale(),
        style: 'long',
        firstDayOfWeek: this.firstDayOfWeek,
      }),
      dayTemplate,
    });
}
```

while we want to have smth like

```js
  _dataTemplate() { // overridable part which has a default implementation
    return dataTemplate;
  }
 __renderData() { // often private, but can be also exposed if more flexibility is needed
    return this.__dataTemplate(this.__data, {
      monthsLabels: getMonthNames({ locale: this.__getLocale() }),
      weekdaysShort: getWeekdayNames({
        locale: this.__getLocale(),
        style: this.weekdayHeaderNotation,
        firstDayOfWeek: this.firstDayOfWeek,
      }),
      weekdays: getWeekdayNames({
        locale: this.__getLocale(),
        style: 'long',
        firstDayOfWeek: this.firstDayOfWeek,
      }),
      dayTemplate,
    });
}
```

PS: Ha-ha, now when I wrote that, feels like we need better terms. Maybe `reusable template` and `bound template` can give more clarity?

**Comment by tlouisse** [comment]
> > '__renderHeader' instead of '_headerTemplate' or 'headerTemplate'. I would opt for '_headerTemplate' (aligned with existing naming convention, one underscore since available for subclassers)
> 
> We never had a situation with reusable templates in the past in `lion`. Calendar is the first component where we have both reusable templates (I suggest to call them just `template functions`) which receive everything via arguments and non-reusable templates bound to the context of a web component (I suggest to call them `render functions`).
> 
> e.g. `__renderData` is a render function which uses the template `dataTemplate`
> 
> https://github.com/ing-bank/lion/pull/37/files#diff-0d8b3c4eac421dbadc591e2ed5b80f6dR280
> 
> Therefore I had to introduce a new convention here. And I don't think it is a good idea to follow the previous approach with no separation between those different things. Or it will be like:
> 
> ```js
>  __dataTemplate() {
>     return dataTemplate(this.__data, {
>       monthsLabels: getMonthNames({ locale: this.__getLocale() }),
>       weekdaysShort: getWeekdayNames({
>         locale: this.__getLocale(),
>         style: this.weekdayHeaderNotation,
>         firstDayOfWeek: this.firstDayOfWeek,
>       }),
>       weekdays: getWeekdayNames({
>         locale: this.__getLocale(),
>         style: 'long',
>         firstDayOfWeek: this.firstDayOfWeek,
>       }),
>       dayTemplate,
>     });
> }
> ```
> 
> while we want to have smth like
> 
> ```js
>   _dataTemplate() { // overridable part which has a default implementation
>     return dataTemplate;
>   }
>  __renderData() { // often private, but can be also exposed if more flexibility is needed
>     return this.__dataTemplate(this.__data, {
>       monthsLabels: getMonthNames({ locale: this.__getLocale() }),
>       weekdaysShort: getWeekdayNames({
>         locale: this.__getLocale(),
>         style: this.weekdayHeaderNotation,
>         firstDayOfWeek: this.firstDayOfWeek,
>       }),
>       weekdays: getWeekdayNames({
>         locale: this.__getLocale(),
>         style: 'long',
>         firstDayOfWeek: this.firstDayOfWeek,
>       }),
>       dayTemplate,
>     });
> }
> ```
> 
> PS: Ha-ha, now when I wrote that, feels like we need better terms. Maybe `reusable template` and `bound template` can give more clarity?

**Comment by tlouisse** [comment]
> We never had a situation with reusable templates in the past in `lion`. Calendar is the first component where we have both reusable templates (I suggest to call them just `template functions`) which receive everything via arguments and non-reusable templates bound to the context of a web component (I suggest to call them `render functions`).

We have all those templates (without underscore) in the field components. They are also part of the input-datepicker now.
Whatever we call them: have a session about them and align on names :)

I will approve the merge request, so that we can continue finishing datepicker.... 👍

**Comment by bashmish** [comment]
@ing-bank/lion can we please make per-line comments to make discussion threads more isolated and manageable?

File: packages/localize/src/date/getMonthNames.js (hunk 2 above)
Comment (bashmish) [reply packages/localize/src/date/getMonthNames.js]: nope

Comment (bashmish) [reply packages/localize/src/date/getMonthNames.js]: it was just not needed

if you look at the implementation of both new functions, you'll see that we can add other `style` options later without changing the API, so let's keep it open for now

**Comment by bashmish** [comment]
> We have all those templates (without underscore) in the field components. They are also part of the input-datepicker now.
> Whatever we call them: have a session about them and align on names :)

Indeed you are right, we have this distinction already in forms.

https://github.com/ing-bank/lion/blob/master/packages/field/src/FormControlMixin.js#L272-L287

Those are templates in the sense I described them: none of them uses the context. So I think we are on the same page here. Let's just document this convention somewhere.

File: packages/localize/src/date/getMonthNames.js (hunk 2 above)
Comment (bashmish) [reply packages/localize/src/date/getMonthNames.js]: @tlouisse please create an issue to implement support for other notations of month names

**Pull request merged**